benchmark/Cajita_FastFourierTransformPerformance.cpp

```diff
@@ -0,0 +1,232 @@
+                {
+                    auto rand = pool.get_state( i + j + k );
+                    lhs_host_view( i, j, k, 0 ) =
+                        Kokkos::rand<decltype( rand ), double>::draw( rand, 0.0,
+                                                                      1.0 );
+                    lhs_host_view( i, j, k, 1 ) =
+                        Kokkos::rand<decltype( rand ), double>::draw( rand, 0.0,
+                                                                      1.0 );
+                }
+
+        // Copy to the device
+        Kokkos::deep_copy( lhs_view, lhs_host_view );
+
+        setup_timer.start( p );
+
+        // Create FFT options
+        Experimental::FastFourierTransformParams params;
+        params.setAllToAll( true );
+        params.setPencils( true );
+        params.setReorder( true );
+
+        auto fft = Experimental::createHeffteFastFourierTransform<
+            double, Device, HostBackendType>( *vector_layout, params );
+
+        setup_timer.stop( p );
+
+        transforms_timer.start( p );
+
+        // Loop over number of runs
+        for ( int t = 0; t < num_runs; ++t )
+        {
+            fft->forward( *lhs, Experimental::FFTScaleFull() );
+            fft->reverse( *lhs, Experimental::FFTScaleNone() );
+        }
+
+        transforms_timer.stop( p );
+    }
+
+    outputResults( stream, "grid_size_per_dim", grid_sizes_per_dim, setup_timer,
+                   comm );
+    outputResults( stream, "grid_size_per_dim", grid_sizes_per_dim,
+                   transforms_timer, comm );
+
+    stream << std::flush;
+}
+
+//---------------------------------------------------------------------------//
+// main
+int main( int argc, char* argv[] )
+{
+    // Initialize environment
+    MPI_Init( &argc, &argv );
+    Kokkos::initialize( argc, argv );
+
+    // Check arguments.
+    if ( argc < 2 )
+        throw std::runtime_error( "Incorrect number of arguments. \n \
+             First argument - file name for output \n \
+             \n \
+             Example: \n \
+             $/: ./FastFourierTransformPerformance test_results.txt\n" );
+
+    // Get the name of the output file.
+    std::string filename = argv[1];
+
+    // Barrier before continuing
+    MPI_Barrier( MPI_COMM_WORLD );
+
+    // Get comm rank;
+    int comm_rank;
+    MPI_Comm_rank( MPI_COMM_WORLD, &comm_rank );
+
+    // Open the output file on rank 0.
+    std::fstream file;
+    if ( 0 == comm_rank )
+        file.open( filename, std::fstream::out );
+
+    // Output file header
+    if ( 0 == comm_rank )
+    {
+        file << "\n";
+        file << "Cajita FFT Performance Benchmark"
+             << "\n";
+        file << "----------------------------------------------"
+             << "\n";
+        file << std::flush;
+    }
+
+    // Run the tests.
+#ifdef Heffte_ENABLE_FFTW
+#ifdef KOKKOS_ENABLE_SERIAL
+    using SerialDevice = Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>;
```

Comment:
If you flip the ifdefs you can reuse the device types

Reply:
Nice! Thanks I switched these